srml/contract/src/lib.rs

```diff
@@ -546,3 +668,103 @@ impl<Gas: As<u64>> Default for Schedule<Gas> {
 		}
 	}
 }
+
+enum RentDecision<T: Trait> {
+	/// Happens when the rent is offset completely by the `rent_deposit_offset`.
+	/// Or Rent has already been payed
+	/// Or account doesn't has rent because no contract or tombstone contract
+	FreeFromRent,
+	/// The contract still have funds to keep itself from eviction. Collect dues.
+	/// It is ensure that account can withdraw dues without dying.
+	CollectDues(BalanceOf<T>),
+	CantWithdrawRentWithoutDying(BalanceOf<T>),
+	AllowanceExceeded(BalanceOf<T>),
+}
+
+impl<T: Trait> Module<T> {
+	// Check rent at this block number
+	fn check_rent(account: &T::AccountId, block_number: T::BlockNumber) -> RentDecision<T> {
+		let contract = match <ContractInfoOf<T>>::get(account) {
+			None | Some(ContractInfo::Tombstone(_)) => return RentDecision::FreeFromRent,
+			Some(ContractInfo::Alive(contract)) => contract,
+		};
+
+		// Rent has already been payed
+		if contract.deduct_block >= block_number {
+			return RentDecision::FreeFromRent
+		}
+
+		let balance = T::Currency::free_balance(account);
+
+		let effective_storage_size = <BalanceOf<T>>::sa(contract.storage_size)
+			.saturating_sub(balance.checked_div(&<Module<T>>::rent_deposit_offset()).unwrap_or(<BalanceOf<T>>::sa(0)));
+
+		let fee_per_block: BalanceOf<T> = effective_storage_size * <Module<T>>::rent_byte_price();
+
+		if fee_per_block.is_zero() {
+			// The rent deposit offset reduced the fee to 0. This means that the contract
+			// gets the rent for free.
+			return RentDecision::FreeFromRent
+		}
+
+		let rent = fee_per_block * <BalanceOf<T>>::sa((block_number.saturating_sub(contract.deduct_block)).as_());
+
+		// Pay rent up to rent_allowance
+		if rent <= contract.rent_allowance {
+			// TODO TODO: this could be one call if ensure_can_withdraw took exitence requirement
+			if balance.saturating_sub(rent) > T::Currency::minimum_balance()
```

Comment:
or we talked on that, hmm it is actually more complicated than I though and my implementation is wrong. in my implementation check_rent can return Cantwithdrawwithoutdying(rent), AllowanceExceeded(rent_allowance)
then on eviction it pays for the amount (rent or rent_allowance depending on case) and then check if superior or equal to subsistance threshold.
(because in my mind an account must be able to pay up to rent allowance, which is false in state-rent-sim)

on the other hand state-rent sim doesn't do so if an account with a balance superior or equal to subsistance threshold gets evicted then (for whatever rent_allowance is) it generates a tombstone. (and pay for rent up to the block he could pay).

Implementing your solution is more difficult because of withdraw. 🤔

Reply:
updated

Reply:
logic has changed a bit for the better I have to double check what I wrote though